tests/oauth.test.js

```diff
@@ -0,0 +1,68 @@
+"use strict";
+
+const got = require("got");
+
+const AppConstants = require("../app-constants");
+const DB = require("../db/DB");
+const getSha1 = require("../sha1-utils");
+const {init, confirmed} = require("../controllers/oauth");
+
+require("./resetDB");
+
+
+jest.mock("got");
+
+
+test("init request sets session cookie and redirects", () => {
+  const mockRequest = { session: { } };
+  const mockResponse = { redirect: jest.fn() };
+
+  init(mockRequest, mockResponse);
+
+  const mockRedirectCallArgs = mockResponse.redirect.mock.calls[0];
+  expect(mockRedirectCallArgs[0]).toMatch(AppConstants.OAUTH_AUTHORIZATION_URI);
+});
+
+
+test("confirmed request checks session cookie, calls FXA for token and email, adds subscriber, and renders", async () => {
+  const testFxAEmail = "fxa@test.com";
+  // Mock the getToken, got, and render calls
+  const mockRequest = { session: { state: { } }, originalUrl: "" };
+  const mockResponse = { render: jest.fn() };
+  const mockFxAClient = { code : { getToken: jest.fn().mockReturnValueOnce({ accessToken: "testToken"}) } };
+  got.mockResolvedValue({ body: `{"email": "${testFxAEmail}"}` });
+
+  await confirmed(mockRequest, mockResponse, mockFxAClient);
+
+  const mockFxACallArgs = mockFxAClient.code.getToken.mock.calls[0];
+  expect(mockFxACallArgs[0]).toBe(mockRequest.originalUrl);
+  expect(mockFxACallArgs[1]).toEqual({state: mockRequest.session.state});
```

Comment:
This is a lot of array access operators. Not sure if additional destructuring would help, or just pollute here. But then again, it's "only" tests, so maybe I'm overthinking.

```js
const [originalUrl, state] = mockFxAClient.code.getToken.mock.calls[0];
expect(originalUrl).toBe(mockRequest.originalUrl);
expect(state).toEqual({state: mockRequest.session.state});

const {verified, email} = (await DB.getSubscribersByHashes([getSha1(testFxAEmail)]))[0];
expect(verified).toBeTruthy();
expect(email).toBe(testFxAEmail);

// etc.
```

Reply:
I like the array de-structuring syntax, but in combination with the jest `mock.calls` it removes the semantics that `calls[0]` is the call.

Can you nest array de-structuring? So ...

```
const [firstCall [originalUrl, state] ] = mockFxAClient.code.getToken.mock;
```

That would be sweet.